pkg/certs/certgen.go

```diff
@@ -0,0 +1,138 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package certs
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/projectcontour/contour/internal/certgen"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// CertgenConfig holds the configuration for the certificate generation process.
+type CertgenConfig struct {
+
+	// KubeConfig is the path to the Kubeconfig file if we're not running in a cluster
+	KubeConfig string
+
+	// Incluster means that we should assume we are running in a Kubernetes cluster and work accordingly.
+	InCluster bool
+
+	// Namespace is the namespace to put any generated config into for YAML or Kube outputs.
+	Namespace string
+
+	// OutputDir stores the directory where any requested files will be output.
+	OutputDir string
+
+	// OutputKube means that the certs generated will be output into a Kubernetes cluster as secrets.
+	OutputKube bool
+
+	// OutputYAML means that the certs generated will be output into Kubernetes secrets as YAML in the current directory.
+	OutputYAML bool
+
+	// OutputPEM means that the certs generated will be output as PEM files in the current directory.
+	OutputPEM bool
+
+	// Lifetime is the number of days for which certificates will be valid.
+	Lifetime uint
+
+	// Overwrite allows certgen to overwrite any existing files or Kubernetes Secrets.
+	Overwrite bool
+
+	// Format specifies how to format the Kubernetes Secrets (must be "legacy" or "compat").
+	Format string
+}
+
+// GenerateCerts performs the actual cert generation steps and then returns the certs for the output function.
+func GenerateCerts(certConfig *CertgenConfig) (map[string][]byte, error) {
```

Comment:
any reason not to combine `GenerateCerts` and `OutputCerts` into a single exported function, since they need to be called one after the other? It'd make for a slightly cleaner public API IMHO.

Reply:
Yeah I'll do that, less API exposed is better. =)

Reply:
In looking into this a bit more, it seems like we should have 3 public methods outlined below. The confusing part about how it's built today is that if you request a cert in `yaml` or `pem` format, you still need to pass a kubeclient which isn't used at all in those two situations. 

1. Kubernetes Secret (`--kube`), `GenerateKubeCertificates(format string, kubeclient *kubernetes.Clientset, overwrite bool)`
2. Yaml (`--yaml`), `GenerateYamlCertificates(format, outputDirectory string, force bool)`
3. PEM (`--pem`), `GeneratePemCertificates(format, outputDirectory string, force bool)`

_Note: We could combine the yaml & pem into a single one, but I'd probably keep separate? /shrug_

This means then that the `doCertgen()` inside `cmd/contour` will look at the flags passed and call each public method appropriately, but external callers have a clean API to use to code against.

Reply:
I propose just one API:
```Go

type CertificateSet struct {
    Contour []byte
    Envoy []byte
    CA []byte
}

func GenerateCertificates(options ...) (CertificateSet, error)
```

The caller can deal with emitting the certificates in the way they need (i.e. contour-operator probably can't directly use any of the specific APIs above). We probably need the options so the caller can do things like set unique SAN entries.

Reply:
The operator can use the apis above as-is. 

The work to serialize them to pem or yaml or secrets is a benefit of using a library as this is intended so the caller doesn't need to reinvent that work.

Reply:
There could be further customizations to add name, etc, etc, but could come as well as needed.

Reply:
> The operator can use the apis above as-is.

Are you sure? The operator doesn't need `GenerateYamlCertificates` or `GenerateYamlCertificates`. It may need `GenerateKubeCertificates`, but (1) depending on the Kubernetes framework it uses, it may not have easy access to the right client struct (2) it may need to specify the name or namespace of the secret (3) it may need to set owner references or other object meta information (4) it may combine the certificates into secrets in way this function does not anticipate.

Reply:
Requirements / needs will change over time, 💯 .

Reply:
I like the sound of @jpeach's suggestion, it seems like a nice, clean API. I think it's probably better to have separate Output functions for each type though, so that we can ensure that things consuming this package output their certs in the right way (ie files have the correct permissions etc, certificates are in the right formats, and so on). Callers don't *have* to use the Output functions, but supplying them makes keeping things in sync easier.